oslo_vmware/rw_handles.py

```diff
@@ -308,10 +313,38 @@ def __init__(self, session, lease, url, file_handle,
 
     def _log_progress(self, progress):
         """Log data transfer progress."""
-        if (progress == 100 or (progress - self._last_logged_progress >=
-                                MIN_PROGRESS_DIFF_TO_LOG)):
-            LOG.debug("Data transfer progress is %d%%.", progress)
-            self._last_logged_progress = progress
+        time_diff = time.time() - self._last_progress_update
+        progress["time_diff"] = datetime.timedelta(seconds=time_diff)
+
+        LOG.info("Data Transfer progress: %(bytes)s"
+                 " of %(total_size)s in %(time_diff)s %(percent)s", progress)
+
+    def _check_progress_stalled(self, progress, now):
+        """Check if we still make progress
+
+        Compares the given progress to the last known progress to see if we
+        still go forward. Emits a warning otherwise.
+        """
+
+        if progress['bytes'] < self._last_logged_progress:
+            LOG.warning('We made negative progress: before %(before)s%%, '
+                        'now %(now)s%%',
+                        {'before': self._last_progress,
+                         'now': progress['percent']})
+            self._last_progress_change = now
+        elif (progress['bytes'] == self._last_logged_progress
+                and (progress['bytes'] != 0 or
+                     now - self._last_progress_change > 300)):
+            time_diff = (now - self._last_progress_change
+                         if self._last_progress_change else 0)
+            LOG.warning('No progress made in %(interval)ss. Currently at '
+                        '%(progress)s%%',
+                        {'interval': datetime.timedelta(seconds=time_diff),
+                         'progress': progress['percent']})
+        else:
+            self._last_progress_change = now
+            self._last_logged_progress = progress['bytes']
+        self._last_progress = progress
```

Comment:
We're re-using `_last_logged_progress` here, which was originally used by `_log_progress()`. We additionally introduce `_last_progress`. Does it make sense to have both? Imho, we should distinguish between the two.